core/src/container.cpp

```diff
@@ -194,6 +194,14 @@ bool ContainerBase::traverseRecursively(const ContainerBase::StageCallback& proc
 	return pimpl()->traverseStages(processor, 1, UINT_MAX);
 }
 
+void ContainerBase::add(Stage::pointer&& stage) {
```

Comment:
I suggest to unify the API of `add()` and `insert()`, i.e. provide an inline forward `add(stage) = insert(stage, -1)`. The error checking, should be moved into the base function `ContainerBase::insert()`.
I'm fine to unify the return value as well, i.e. switching to void and just throw?

Reply:
If we move away from the `bool` return values, `insert` does not mirror the stl container interface anymore (although `insert` is astonishingly not defined in a named concept as far as I know). Aside from being a standard container, I have never seen a reason for this function to begin with, as users should never insert stages anywhere but at the end.

Would it be sufficient for you if I add another check for a valid stage inside `insert` (which should have been there either way)?
I would like to keep the distinction between `bool` return and exceptions and also the one between "invalid pointer passed" and "insert operation failed".

Reply:
In which fashion do you want to distinguish these two error cases? Currently, they can only be distinguished by their error message, but not programmatically (both throw a runtime error).
Adding an additional check for stage validity in `insert` makes sense. Do you plan to return false or throw in this case? A reasonable distinction between insert and add could be as follows:
* `insert` is considered "low-level", not throwing, but only returning false in case of an error.
* `add` is considered "high-level", throwing (the current) exceptions on error.

> ... users should never insert stages anywhere but at the end.

I disagree. Considering preconfigured serial containers, e.g. for grasping, it is useful to be able to inject further stages, e.g. to modify the ACM.

Reply:
> In which fashion do you want to distinguish these two error cases?

By error message, as you say. Programmatically, there is no reason to discriminate between them as user code could check the pointer before calling. Telling the programmer his pointer is invalid does have merit though. :)

As [`setParent` already warns using `ROS_ERROR`](https://github.com/ros-planning/moveit_task_constructor/pull/149/files#diff-e6edf6dc18e6748048d781f47155c8c6R122), I moved the message to another `ROS_ERROR` log in `insert` now.

The `add`/`insert` distinction you give is what I had in mind, yes.
If you just put together a bunch of add calls, you do not want to check every single return value that way. At the same time, if you need programmatic flexibility, you can use `insert` instead.

-- 

> > ... users should never insert stages anywhere but at the end.
> 
> I disagree.

I know :grin: . I guess we just accept our difference in opinion here, as long as both use-cases are covered.